src/tls.py

```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling Kafka TLS configuration."""
+
+import logging
+import os
+import socket
+import subprocess
+from typing import List, Optional
+
+from charms.kafka.v0.kafka_snap import SNAP_CONFIG_PATH
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    TLSCertificatesRequiresV1,
+    generate_csr,
+    generate_private_key,
+)
+from ops.charm import ActionEvent, RelationJoinedEvent
+from ops.framework import Object
+from ops.model import Relation
+
+from literals import TLS_RELATION
+from utils import generate_password, parse_tls_file, safe_write_to_file
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaTLS(Object):
+    """Handler for managing the client and unit TLS keys/certs."""
+
+    def __init__(self, charm):
+        super().__init__(charm, "tls")
+        self.charm = charm
+        self.certificates = TLSCertificatesRequiresV1(self.charm, TLS_RELATION)
+
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_created, self._on_certificates_created
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_joined, self._on_certificates_joined
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_broken, self._on_certificates_broken
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring, self._on_certificate_expiring
+        )
+        self.framework.observe(self.charm.on.set_tls_private_key_action, self._set_tls_private_key)
+
+    def _on_certificates_created(self, _):
+        """Handler for `certificates_relation_created` event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        self.peer_relation.data[self.charm.app].update({"tls": "enabled"})
```

Comment:
TLS won't be enabled until the certificate is available and the applications restart

Reply:
This is more of a notification for the charm itself. Adding this key will trigger a `peer_relation_changed` so we can change properties, check if zk is using tls as well...